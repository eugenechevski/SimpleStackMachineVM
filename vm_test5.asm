	# $Id: vm_test5.asm,v 1.4 2024/08/29 20:26:00 leavens Exp $
	.text start
start:	ADDI $sp, 0, 0x7  # stack top is 7
	ADDI $sp, -1, 0xf # 1 above stack top is 15
	ARI $sp, -1       # make the stack top be the location with 15
	LWR $r4, $gp, 0    # $r4 should be 2
	AND $sp, 0, $sp, 0 # stack top should (still) be 15
	SLL  $sp, 0, 5    # stack top should be 480
	ADDI $sp, 0, -405 # stack top is 75
        SRI $sp, 1        # allocate a word on the stack
	ADDI $sp, 0, 2    # put 2 on top of stack
	SLL $sp, 0, 7     # top of stack is 256
	SRL $sp, 0, 2     # stack top is 64
        ARI $sp, -1       # allocate a word on the stack
	LIT $sp, 0, 11    # put 11 on top of stack
        LWR $r5, $sp, 0   # put 11 in $r5
	XOR $sp, 0, $sp, 0 # make top of stack be 0
	SWR $sp, 0, $r5   # put 11 on top of stack
	NEG $sp, 0, $sp, 0 # make top of stack be -11
        LIT $sp, 1, 75    # put 75 in memory[SP+1]
        ARI $sp, 1        # pop the stack
	BNE $sp, 1, 3     # skip if top 2 stack elements are not equal
        CALL y
	JMPA done
	CALL n
	JMPA bad
done:	EXIT 0
bad:	EXIT 1

y:	NOTR
	PSTR $gp, 5
	STRA
	RTN

n:	NOTR
	PSTR $gp, 1
	STRA
	RTN

	.data 512
	WORD TWO = 2
	STRING[1] No = "No\n"
	WORD L = 76
	WORD H = 72
	WORD nl = 10
	STRING[2] Yes = "Yes!\n"
	.stack 4096
	.end
	
