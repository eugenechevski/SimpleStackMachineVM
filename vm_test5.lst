Address Instruction
     0: ADDI $sp, 0, 7
     1: ADDI $sp, -1, 15
     2: ARI $sp, -1
     3: LWR $r4, $gp, 0
     4: AND $sp, 0, $sp, 0
     5: SLL $sp, 0, 5
     6: ADDI $sp, 0, -405
     7: SRI $sp, 1
     8: ADDI $sp, 0, 2
     9: SLL $sp, 0, 7
    10: SRL $sp, 0, 2
    11: ARI $sp, -1
    12: LIT $sp, 0, 11
    13: LWR $r5, $sp, 0
    14: XOR $sp, 0, $sp, 0
    15: SWR $sp, 0, $r5
    16: NEG $sp, 0, $sp, 0
    17: LIT $sp, 1, 75
    18: ARI $sp, 1
    19: BNE $sp, 1, 3	# target is word address 22
    20: CALL 26	# target is word address 26
    21: JMPA 24	# target is word address 24
    22: CALL 30	# target is word address 30
    23: JMPA 25	# target is word address 25
    24: EXIT 0
    25: EXIT 1
    26: NOTR 
    27: PSTR $gp, 5
    28: STRA 
    29: RTN 
    30: NOTR 
    31: PSTR $gp, 1
    32: STRA 
    33: RTN 
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

