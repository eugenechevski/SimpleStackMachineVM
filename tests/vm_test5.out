      PC: 0
GPR[$gp]: 512  	GPR[$sp]: 4096 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

    4096: 0	

==>      0: ADDI $sp, 0, 7
      PC: 1
GPR[$gp]: 512  	GPR[$sp]: 4096 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

    4096: 7	

==>      1: ADDI $sp, -1, 15
      PC: 2
GPR[$gp]: 512  	GPR[$sp]: 4096 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     
    4095: 15	
    4096: 7	

==>      2: ARI $sp, -1
      PC: 3
GPR[$gp]: 512  	GPR[$sp]: 4095 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

    4095: 15	    4096: 7	

==>      3: LWR $r4, $gp, 0
      PC: 4
GPR[$gp]: 512  	GPR[$sp]: 4095 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

    4095: 15	    4096: 7	

==>      4: AND $sp, 0, $sp, 0
      PC: 5
GPR[$gp]: 512  	GPR[$sp]: 4095 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

    4095: 15	    4096: 7	

==>      5: SLL $sp, 0, 5
      PC: 6
GPR[$gp]: 512  	GPR[$sp]: 4095 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

    4095: 480	    4096: 7	

==>      6: ADDI $sp, 0, -405
      PC: 7
GPR[$gp]: 512  	GPR[$sp]: 4095 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

    4095: 75	    4096: 7	

==>      7: SRI $sp, 1
      PC: 8
GPR[$gp]: 512  	GPR[$sp]: 4094 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

    4094: 0	    4095: 75	    4096: 7	

==>      8: ADDI $sp, 0, 2
      PC: 9
GPR[$gp]: 512  	GPR[$sp]: 4094 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

    4094: 2	    4095: 75	    4096: 7	

==>      9: SLL $sp, 0, 7
      PC: 10
GPR[$gp]: 512  	GPR[$sp]: 4094 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

    4094: 256	    4095: 75	    4096: 7	

==>     10: SRL $sp, 0, 2
      PC: 11
GPR[$gp]: 512  	GPR[$sp]: 4094 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

    4094: 64	    4095: 75	    4096: 7	

==>     11: ARI $sp, -1
      PC: 12
GPR[$gp]: 512  	GPR[$sp]: 4093 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

    4093: 0	    4094: 64	    4095: 75	    4096: 7	

==>     12: LIT $sp, 0, 11
      PC: 13
GPR[$gp]: 512  	GPR[$sp]: 4093 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

    4093: 11	    4094: 64	    4095: 75	    4096: 7	

==>     13: LWR $r5, $sp, 0
      PC: 14
GPR[$gp]: 512  	GPR[$sp]: 4093 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 11   	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

    4093: 11	    4094: 64	    4095: 75	    4096: 7	

==>     14: XOR $sp, 0, $sp, 0
      PC: 15
GPR[$gp]: 512  	GPR[$sp]: 4093 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 11   	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

    4093: 0	    4094: 64	    4095: 75	    4096: 7	

==>     15: SWR $sp, 0, $r5
      PC: 16
GPR[$gp]: 512  	GPR[$sp]: 4093 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 11   	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

    4093: 11	    4094: 64	    4095: 75	    4096: 7	

==>     16: NEG $sp, 0, $sp, 0
      PC: 17
GPR[$gp]: 512  	GPR[$sp]: 4093 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 11   	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

    4093: -11	    4094: 64	    4095: 75	    4096: 7	

==>     17: LIT $sp, 1, 75
      PC: 18
GPR[$gp]: 512  	GPR[$sp]: 4093 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 11   	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     

    4093: -11	    4094: 75	    4095: 75	    4096: 7	

==>     18: ARI $sp, 1
      PC: 19
GPR[$gp]: 512  	GPR[$sp]: 4094 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 11   	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     
    4093: -11	
    4094: 75	    4095: 75	    4096: 7	

==>     19: BNE $sp, 1, 3	# target is word address 22
      PC: 20
GPR[$gp]: 512  	GPR[$sp]: 4094 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 11   	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     
    4093: -11	
    4094: 75	    4095: 75	    4096: 7	

==>     20: CALL 26	# target is word address 26
      PC: 26
GPR[$gp]: 512  	GPR[$sp]: 4094 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 11   	GPR[$r6]: 0    	GPR[$ra]: 21   
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     
    4093: -11	
    4094: 75	    4095: 75	    4096: 7	

==>     26: NOTR 
Yes!
      PC: 29
GPR[$gp]: 512  	GPR[$sp]: 4094 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 11   	GPR[$r6]: 0    	GPR[$ra]: 21   
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     
    4093: -11	
    4094: 5	    4095: 75	    4096: 7	

==>     29: RTN 
      PC: 21
GPR[$gp]: 512  	GPR[$sp]: 4094 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 11   	GPR[$r6]: 0    	GPR[$ra]: 21   
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     
    4093: -11	
    4094: 5	    4095: 75	    4096: 7	

==>     21: JMPA 24	# target is word address 24
      PC: 24
GPR[$gp]: 512  	GPR[$sp]: 4094 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 2    
GPR[$r5]: 11   	GPR[$r6]: 0    	GPR[$ra]: 21   
     512: 2	     513: 683854	     514: 76	     515: 72	     516: 10	
     517: 561210713	     518: 10	     519: 0	        ...     
    4093: -11	
    4094: 5	    4095: 75	    4096: 7	

==>     24: EXIT 0
